[SPARK-14050][ML] Add multiple languages support and additional methods for Stop Words Remover

burakkose opened the pull request:
Apache Spark is a global engine, so It should appeal to everyone as much as possible. I added multiple language support for StopWordRemover, and used nltk's language list except for English(English is still same). I added some facilitator functions such as setLanguage, setAdditionalWords and setIgnoredWords. 

English is the default. If you want to change the language, use setLanguage. For example; setLanguage("danish").

burakkose committed:
add language files

burakkose committed:
add multi-language support for stop words

burakkose committed:
add new tests for StopWordsRemover

burakkose committed:
adjust resource files

burakkose committed:
adjust resource files

burakkose committed:
fix stopwords bug

burakkose committed:
update comment lines

AmplabJenkins commented:
Can one of the admins verify this patch?

mengxr commented:
@burakkose I guess it is okay to copy the lists from NLTK since it is Apache licensed. Could you add a header to each stopword file and put a link there? Thanks!

mengxr commented:
ok to test

mengxr commented on mllib/src/main/resources/org/apache/spark/ml/feature/stopwords/english.txt:
```diff
@@ -0,0 +1,319 @@
+a
```
If other lists are from NLTK, maybe we should use their English stopwords too. It would be good to make quick comparison and check the differences.

SparkQA commented:
**[Test build #53745 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53745/consoleFull)** for PR 11871 at commit [`6d215b3`](https://github.com/apache/spark/commit/6d215b31a205c4a79e8cc0ef6963d239941e80ff).

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -31,51 +31,20 @@ import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
+    "hungarian", "italian", "norwegian", "portuguese", "russian", "spanish", "swedish", "turkish")
+
+  /** Languages and stopwords map */
+  val languageMap = supportedLanguages.map{
```
Loading all lists by default is unnecessary. Check supported languages in `readStopWords` load there.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -31,51 +31,20 @@ import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
+  }
+
+  /** Supported languages list must be lowercase */
+  val supportedLanguages = Array("danish", "dutch", "english", "finnish", "french", "german",
```
`private val supportedLanguages: Set[String] = Set(...)`. This should be a private val and using `Set` to test membership.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -31,51 +31,20 @@ import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
-    "within", "without", "would", "yet", "you", "your", "yours", "yourself", "yourselves")
+  /** Read stop words list from resources */
+  def readStopWords(language: String): Array[String] = {
+    val is = getClass.getResourceAsStream(s"/org/apache/spark/ml/feature/stopwords/$language.txt")
```
Validate language before loading resource file.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -98,13 +67,16 @@ class StopWordsRemover(override val uid: String)
   /**
    * the stop words set to be filtered out
-   * Default: [[StopWords.English]]
+   * Default: [[StopWords.languageMap("english")]]
```
It won't show up correctly since `object StopWords` is private. `StopWords.English` is not correct either. We can just mention that by default it is unset and it will be set to the default list for the selected language during transformation.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -98,13 +67,16 @@ class StopWordsRemover(override val uid: String)
-  def setStopWords(value: Array[String]): this.type = set(stopWords, value)
+  def setStopWords(value: Array[String]): this.type = {
+    set(stopWords, value)
+    set(language, "unknown")
```
We shouldn't set other parameters in the setters since their are other ways to overwrite param values. Let's keep the default language to `english` and if `stopWords` is set, ignore the language setting. This logic should happen in `transform`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -123,21 +95,74 @@ class StopWordsRemover(override val uid: String)
+    * Default: "english"
+    * @group param
+    */
+  val language: Param[String] = new Param[String](this, "language", "stopwords language")
```
In the doc, we should mention supported languages.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -123,21 +95,74 @@ class StopWordsRemover(override val uid: String)
+  /** @group setParam */
+  def setLanguage(value: String): this.type = {
+    val lang = value.toLowerCase
+    require(StopWords.languageMap.contains(lang), s"$lang is not in language list")
```
use ParamValidators.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -123,21 +95,74 @@ class StopWordsRemover(override val uid: String)
+    val lang = value.toLowerCase
+    require(StopWords.languageMap.contains(lang), s"$lang is not in language list")
+    set(language, lang)
+    set(stopWords, StopWords.languageMap(lang))
```
Ditto. Do not set the stop words here. Implement the logic in `transform` instead.

mengxr commented:
@burakkose I made a quick pass. Just want to mention another option for the implementation. Instead of having `language`, `ignoredWords`, and `additionalWords`, we can separate the lists from `StopwordsRemover`:

~~~scala
val stopWords = StopWordsRemover.loadStopWords("turkish").toSet ++ Set("a") -- Set("b"))
val swr = new StopWordsRemover()
  .setStopWords(stopWords.toArray)
...
~~~

This makes more composite code.

SparkQA commented:
**[Test build #53745 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53745/consoleFull)** for PR 11871 at commit [`6d215b3`](https://github.com/apache/spark/commit/6d215b31a205c4a79e8cc0ef6963d239941e80ff).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53745/
Test FAILed.

burakkose commented:
@mengxr , thank you for quick review, I am working on your comment, this will be my first commit, so if you review my new changes, I am happy.

srowen commented:
@burakkose the `LICENSE` file and other license info probably needs updating if you're including this. I can help if you'll point me to the source of this data and its license.

burakkose commented:
@srowen , I got from http://www.nltk.org/nltk_data/ , Stopwords Corpus, and they mentioned that 
"
They were obtained from:
http://anoncvs.postgresql.org/cvsweb.cgi/pgsql/src/backend/snowball/stopwords/
The English list has been augmented
https://github.com/nltk/nltk_data/issues/22
"

burakkose committed:
update stop words list

burakkose committed:
update stopwordsremover

SparkQA commented:
**[Test build #53784 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53784/consoleFull)** for PR 11871 at commit [`41cd258`](https://github.com/apache/spark/commit/41cd25815af3baa8fe9ed9336812f436d7ed7bd5).

SparkQA commented:
**[Test build #53784 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53784/consoleFull)** for PR 11871 at commit [`41cd258`](https://github.com/apache/spark/commit/41cd25815af3baa8fe9ed9336812f436d7ed7bd5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53784/
Test FAILed.

burakkose committed:
fix test case bug

After updating English stop words list, "d" is a stop word.

SparkQA commented:
**[Test build #53788 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53788/consoleFull)** for PR 11871 at commit [`4d1812a`](https://github.com/apache/spark/commit/4d1812aae64b0b15312940b1a6c42e19f9686480).

burakkose commented:
@mengxr , can you review again? However, I have a problem about charset. What is the best solution for "java.nio.charset.MalformedInputException: Input length = 1".

SparkQA commented:
**[Test build #53788 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53788/consoleFull)** for PR 11871 at commit [`4d1812a`](https://github.com/apache/spark/commit/4d1812aae64b0b15312940b1a6c42e19f9686480).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53788/
Test FAILed.

burakkose committed:
fix encoding

SparkQA commented:
**[Test build #53822 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53822/consoleFull)** for PR 11871 at commit [`a308622`](https://github.com/apache/spark/commit/a30862231c3944c55c96cc94e162f61614aee6d5).

SparkQA commented:
**[Test build #53822 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53822/consoleFull)** for PR 11871 at commit [`a308622`](https://github.com/apache/spark/commit/a30862231c3944c55c96cc94e162f61614aee6d5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53822/
Test FAILed.

burakkose committed:
fix pyspark test

SparkQA commented:
**[Test build #53868 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53868/consoleFull)** for PR 11871 at commit [`2e7c54e`](https://github.com/apache/spark/commit/2e7c54e5c17e7c5672a43ffc28acb207e94bf28a).

SparkQA commented:
**[Test build #53868 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53868/consoleFull)** for PR 11871 at commit [`2e7c54e`](https://github.com/apache/spark/commit/2e7c54e5c17e7c5672a43ffc28acb207e94bf28a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53868/
Test PASSed.

srowen commented:
OK, the license for the stopwords is OK. It's the PostgreSQL license, which is BSD-like. You will want to add an entry for PostgresSQL licenses in the style that you see other BSD licenses recorded in the `LICENSE` file, and then add a copy of http://anoncvs.postgresql.org/cvsweb.cgi/pgsql/COPYRIGHT at `licenses/LICENSE-postgresql.txt`

burakkose committed:
add licence for stop words list

burakkose commented:
@srowen , thank you for your help, I added the licence.

SparkQA commented:
**[Test build #53946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53946/consoleFull)** for PR 11871 at commit [`7efda40`](https://github.com/apache/spark/commit/7efda40e39663deef0b0884a7bfca13b5d10d706).

SparkQA commented:
**[Test build #53946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53946/consoleFull)** for PR 11871 at commit [`7efda40`](https://github.com/apache/spark/commit/7efda40e39663deef0b0884a7bfca13b5d10d706).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53946/
Test FAILed.

burakkose commented:
Jenkins, retest this please

burakkose commented:
Jenkins, retest this please.

burakkose commented:
retest this please

srowen commented on licenses/LICENCE-postgresql.txt:
```diff
@@ -0,0 +1,24 @@
+PostgreSQL Database Management System
```
Yes, though nit: "LICENSE" vs "LICENCE". We're using the US spelling consistently

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54028 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54028/consoleFull)** for PR 11871 at commit [`7efda40`](https://github.com/apache/spark/commit/7efda40e39663deef0b0884a7bfca13b5d10d706).

SparkQA commented:
**[Test build #54028 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54028/consoleFull)** for PR 11871 at commit [`7efda40`](https://github.com/apache/spark/commit/7efda40e39663deef0b0884a7bfca13b5d10d706).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54028/
Test PASSed.

burakkose committed:
change licence to license

SparkQA commented:
**[Test build #54058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54058/consoleFull)** for PR 11871 at commit [`a066e8b`](https://github.com/apache/spark/commit/a066e8b34ec4824fa26a1e306e197b66400f5ccb).

SparkQA commented:
**[Test build #54058 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54058/consoleFull)** for PR 11871 at commit [`a066e8b`](https://github.com/apache/spark/commit/a066e8b34ec4824fa26a1e306e197b66400f5ccb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54058/
Test FAILed.

burakkose commented:
@srowen, spelling's ok, need re-test, does this pr need anything else?

mengxr commented:
Shall we put a README.txt under `mllib/src/main/resources/org/apache/spark/ml/feature/stopwords/` and mention the source of stop words?

burakkose commented:
Do we need that? In my view, stop words list will not be updated anymore, and we have already added the license.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -31,51 +31,16 @@ import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
  */
 private[spark] object StopWords {
```
This class is package private and hence users do not have access to `readStopWords`. We can either make it public or move the methods to `object StopWordsRemover`. If we take the former approach, we should check Java compatibility. I like the latter approach better.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -31,51 +31,16 @@ import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
-    "wherein", "whereupon", "wherever", "whether", "which", "while", "whither",
-    "who", "whoever", "whole", "whom", "whose", "why", "will", "with",
-    "within", "without", "would", "yet", "you", "your", "yours", "yourself", "yourselves")
+  /** Read stop words list from resources */
```
List supported languages. Or make `supportedLanguages` public and link to it.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -31,51 +31,16 @@ import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
-    "who", "whoever", "whole", "whom", "whose", "why", "will", "with",
-    "within", "without", "would", "yet", "you", "your", "yours", "yourself", "yourselves")
+  /** Read stop words list from resources */
+  def readStopWords(language: String): Array[String] = {
```
`loadStopWords` might be a better name here.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -31,51 +31,16 @@ import org.apache.spark.sql.types.{ArrayType, StringType, StructType}
+  }
+
+  /** Supported languages list must be lowercase */
+  private val supportedLanguages = Set("danish", "dutch", "english", "finnish", "french", "german",
```
If we make it public, we need to return `Array[String]` instead of `Set[String]` for Java compatibility.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -123,21 +88,43 @@ class StopWordsRemover(override val uid: String)
+   * Default: "English"
+   * @group param
+   */
+  val language: Param[String] = new Param[String](this, "language", "stopwords language")
```
I guess we don't need `language` as a param anymore.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -160,4 +147,11 @@ object StopWordsRemover extends DefaultParamsReadable[StopWordsRemover] {
+
+  /**
+   * Stop words for the language
+   * Supported languages: Danish, Dutch, English, Finnish, French, German, Hungarian,
```
We use lowercase in MLlib for string values, so `"danish"`, `"dutch"`, etc.

mengxr commented:
@burakkose We can merge `StopWords` into `StopWordsRemover` to simplify the implementation. I see the `language` param provides a convenient way to set parameters, but eventually we need to add it in Python as well and document the behavior when both `language` and `stopWords` are set, just more maintenance cost. So I'd recommend removing it for now.

burakkose commented:
@mengxr, Alright then, these are my plans.

I am going to merge `StopWords` with `StopWordsRemover`(delete readStopWords), and we will read stopwords list from `loadStopWords` in object `StopWordsRemover`. So If users want to load stopwords directly, they will do that.

`val stopWords = StopWordsRemover.loadStopWords("english").toSet`

But I have a question about `language` param. If we don't keep the param, how should they do this?

```scala
val remover = new StopWordsRemover()
      .setInputCol("raw")
      .setOutputCol("filtered")
      .setLanguage("danish")
```
Should we force to do that?

```scala
 val stopWords = StopWordsRemover.loadStopWords("danish").toSet ++ Set("python", "scala")
    val remover = new StopWordsRemover()
      .setInputCol("raw")
      .setOutputCol("filtered")
      .setStopWords(stopWords.toArray)
```

mengxr commented:
~~~scala
    val remover = new StopWordsRemover()
      .setInputCol("raw")
      .setOutputCol("filtered")
      .setStopWords(StopWordsRemover.loadStopWords("danish"))
~~~

It doesn't complicate the code by much but the interface is cleaner and it is slightly easier for us to maintain the Python API (just adding a class method to `StopWordsRemover` in Python).

burakkose committed:
add readme for stopwords list

burakkose committed:
merge StopWords into StopWordsRemover

burakkose committed:
add python stopwords support for language selection

burakkose committed:
add new tests for stopwords

SparkQA commented:
**[Test build #54184 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54184/consoleFull)** for PR 11871 at commit [`789342f`](https://github.com/apache/spark/commit/789342f2d26759db180868a9f59b02c8f85cc835).

SparkQA commented:
**[Test build #54184 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54184/consoleFull)** for PR 11871 at commit [`789342f`](https://github.com/apache/spark/commit/789342f2d26759db180868a9f59b02c8f85cc835).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54184/
Test FAILed.

burakkose commented:
@mengxr , you will like this time :) , added a static method to Python,readme for resources, deleted `StopWords` and `language` . But we need to retest.

burakkose commented:
Jenkins retest this please

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #54446 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54446/consoleFull)** for PR 11871 at commit [`789342f`](https://github.com/apache/spark/commit/789342f2d26759db180868a9f59b02c8f85cc835).

SparkQA commented:
**[Test build #54446 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54446/consoleFull)** for PR 11871 at commit [`789342f`](https://github.com/apache/spark/commit/789342f2d26759db180868a9f59b02c8f85cc835).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54446/
Test PASSed.

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -123,21 +71,26 @@ class StopWordsRemover(override val uid: String)
-          terms.filter(s => !lowerStopWords.contains(toLower(s)))
-        }
+      udf { terms: Seq[String] =>
+        terms.filter(s => !stopWordsSet.contains(s))
```
I think this can be `terms.filterNot(stopWordsSet.contains)`?
It seems like this code path will always pay the cost of making a set out of the stopwords. It's not huge but wonder if it makes sense to store a ref to the set once?

burakkose replied:
Can you give more information about that case. What is the best way for you?

srowen replied:
Can you save a reference to the active set of stopwords instead of making the list into a set each time? might be more natural to have a defensive copy anyway.

burakkose replied:
Yes, I will fix. Do you have any additional suggestions about the pull-request, such as additional features?

srowen replied:
See question below about null words

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -123,21 +71,26 @@ class StopWordsRemover(override val uid: String)
+      }
+    } else {
+      val toLower = (s: String) => if (s != null) s.toLowerCase else s
+      val lowerStopWords = stopWordsSet.map(toLower(_)).toSet
```
When would the stopwords set ever have a null?

burakkose replied:
Before my editing, they wrote that condition. I thought as you said. However, user may do that.
```
//Other operations to assign to the word. Just an example
val word: String = null
val stopwords = Array(word)
val remover = new StopWordsRemover()
      .setInputCol("raw")
      .setOutputCol("filtered")
      .setStopWords(stopWords)
```

srowen replied:
OK, if we don't treat that as an error, then null can be filtered when the stopwords set is created. In fact it can be lowercased at that time too.

burakkose replied:
Sorry for late response, I have midterms in these days, I will work on these isues as soon as possible.

mengxr commented:
@burakkose There were some merge conflicts introduced by recent commits. So please rebase master when you update this PR. Thanks!

holdenk commented on mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala:
```diff
@@ -98,7 +46,7 @@ class StopWordsRemover(override val uid: String)
   /**
    * the stop words set to be filtered out
-   * Default: [[StopWords.English]]
+   * Default: [[Array.empty]]
```
This could be a little clearer with the scaladoc, I think we should mention that Array.empty actually implies loading the english stop words. (Or we could just have the default be the loaded version of the english stop words as is done in the PySpark code).

mengxr commented:
I"m going to send a PR based on this. So we can catch 2.0.

burakkose commented:
@mengxr , I couldn't find free time, sorry for that. I actually wrote new codes, and I was just waiting for tests. I am going to send a new PR.

mengxr commented:
@burakkose I sent out a PR at https://github.com/apache/spark/pull/12843 and it would be great if you can help review it. I think we should get this one into Spark 2.0. There is also a TODO to add locale support. If you have time, could you start working on https://issues.apache.org/jira/browse/SPARK-15064? Thanks!

burakkose committed:
code review and locale support

burakkose committed:
Merge remote-tracking branch 'upstream/master' into StopWordsImp

# Conflicts:
#	mllib/src/main/scala/org/apache/spark/ml/feature/StopWordsRemover.scala
#	mllib/src/test/scala/org/apache/spark/ml/feature/StopWordsRemoverSuite.scala
#	python/pyspark/ml/feature.py
#	python/pyspark/ml/tests.py

SparkQA commented:
**[Test build #57686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57686/consoleFull)** for PR 11871 at commit [`bca7c01`](https://github.com/apache/spark/commit/bca7c0112dce12526d9539d7a6f96325a4c8cc39).

SparkQA commented:
**[Test build #57686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57686/consoleFull)** for PR 11871 at commit [`bca7c01`](https://github.com/apache/spark/commit/bca7c0112dce12526d9539d7a6f96325a4c8cc39).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57686/
Test FAILed.

burakkose committed:
fix stylecheck

SparkQA commented:
**[Test build #57689 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57689/consoleFull)** for PR 11871 at commit [`cb786ee`](https://github.com/apache/spark/commit/cb786eef0f75aa19d9416ed1c07b90510b4bf70b).

SparkQA commented:
**[Test build #57689 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57689/consoleFull)** for PR 11871 at commit [`cb786ee`](https://github.com/apache/spark/commit/cb786eef0f75aa19d9416ed1c07b90510b4bf70b).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57689/
Test FAILed.

burakkose commented:
@mengxr, can you check? I added locale support.

burakkose committed:
address feedback

SparkQA commented:
**[Test build #57785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57785/consoleFull)** for PR 11871 at commit [`01471ec`](https://github.com/apache/spark/commit/01471ec2a74ff86dfaa417509d0f90e2db80b768).

burakkose commented:
@mengxr, can you check? I added the locale support, and applied your changes. I haven't opened a new pull request for the locale support.

SparkQA commented:
**[Test build #57785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57785/consoleFull)** for PR 11871 at commit [`01471ec`](https://github.com/apache/spark/commit/01471ec2a74ff86dfaa417509d0f90e2db80b768).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57785/
Test FAILed.

burakkose committed:
fix locale

SparkQA commented:
**[Test build #57806 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57806/consoleFull)** for PR 11871 at commit [`dec0634`](https://github.com/apache/spark/commit/dec0634a574124ab53c706b14982a6c81a282c97).

SparkQA commented:
**[Test build #57806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57806/consoleFull)** for PR 11871 at commit [`dec0634`](https://github.com/apache/spark/commit/dec0634a574124ab53c706b14982a6c81a282c97).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57806/
Test FAILed.

Halawa13 commented:
Hello,

I'm new to programming pyspark, i have problem with this code 

**from pyspark.ml.feature import StopWordsRemover
df = sc.createDataFrame([(0,["je", "suis", "malade", "comme", "la", "dernierer"]),(1,["si", "non", "tu", "vas", "bien"])],["label", "raw"])
remover = StopWordsRemover(inputCol="raw", outputCol="filtered")
remover.transform(sentenceData).show(truncate=False)**

I want use "loadDefaultStopWords("french")" bat i don't now how use it.
I tried 
remover.loadDefaultStopWords("french").transform(sentenceData).show(truncate=False) : 
It is not working
